### PR TITLE
[GPU] Onednn first conv

### PR DESCRIPTION
**Pull request by byungilm**
### Details:
 - *item1*
 - *...*

### Tickets:
 - 73925
 - 75081
 - 74864

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/graph_optimizer/add_required_reorders.cpp [hunk 1]
```diff
@@ -41,6 +41,8 @@ void add_required_reorders::add_reorder(program& p, program_node* node, program_
         throw std::runtime_error("Internal Error: container index out of range exception.");
     }
     p.add_intermediate(new_reorder_node, *usr, idx);
+
+    usr->get_dependency(idx).set_output_layout(reorder_layout, false);
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/add_required_reorders.cpp:45]: * Is it really required fix here? Why is the layout inconsistent?
* If it is required patch, please add a comment.
* If it is required patch, please test with dlbenchmark on previous generation of GPU.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp [hunk 2]
```diff
@@ -386,9 +385,7 @@ void remove_redundant_reorders::run(program& p) {
         if (usr->as<convolution>().get_preferred_impl_type() != impl_types::onednn &&
             usr->get_output_layout().format != format::fs_b_yx_fsv32)
             return false;
-        if (usr->as<convolution>().get_preferred_impl_type() == impl_types::onednn &&
-            (usr->get_output_layout().format != format::b_fs_yx_fsv32 ||
-            !lo.needs_onednn_bfyx_to_blocked(dep_layout.format, usr->get_output_layout().format, dep_layout, usr->as<convolution>())))
+        if (usr->as<convolution>().get_preferred_impl_type() == impl_types::onednn)
             return false;
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp:389]: nit: what about like below?
```
        if (usr->as<convolution>().get_preferred_impl_type() == impl_types::onednn)
            return false;
        if (usr->get_output_layout().format != format::fs_b_yx_fsv32)
            return false;
```

File: src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp [hunk 3]
```diff
@@ -578,6 +578,27 @@ void reorder_inputs::run(program& p, layout_optimizer& lo, reorder_factory& rf)
             }
         }
 
+        auto& input = conv_node.input();
+        if (conv_node.impl_type == impl_types::onednn && input.get_dependencies().front()->get_output_layout().format == format::bfyx) {
+            // For supporting optimizied onednn first conv, the input format from prev reorder to this conv is changed to a recommended format by onednn.
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp:583]: nit: `optimizied` -> `optimized`

File: src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp [hunk 4]
```diff
@@ -578,6 +578,27 @@ void reorder_inputs::run(program& p, layout_optimizer& lo, reorder_factory& rf)
...
+            auto input_layout = input.get_output_layout();
+            if (lo.needs_onednn_bfyx_to_blocked(format::bfyx, conv_node.get_output_layout().format, input_layout, conv_node)) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp:585]: Now we need to change the name of this function. It is not `bfyx_to_blocked` anymore.
(and I think we need to update the contents of the function itself, too)

File: src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp [hunk 5]
```diff
@@ -578,6 +578,27 @@ void reorder_inputs::run(program& p, layout_optimizer& lo, reorder_factory& rf)
...
+                auto new_layout = input_layout;
+                if (new_layout.data_type == data_types::f16)
+                    new_layout.format = format::bs_fs_yx_bsv8_fsv2;
+                else if (new_layout.data_type == data_types::u8 || new_layout.data_type == data_types::i8)
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp:589]: `data_type_traits::is_i8_u8(new_layout.data_type)`

File: src/plugins/intel_gpu/src/graph/include/layout_optimizer.h [hunk 6]
```diff
@@ -209,5 +209,8 @@ class layout_optimizer {
     /// As an example, if a reorder has 2 reorder users that each reorder has a convolution user, then merge is not done when those 2 convolutions
     /// have different result from this function.
     bool needs_onednn_bfyx_to_blocked(format fmt_prev, format fmt_next, layout& prev_output_layout, const convolution_node& node);
+    /// @brief Validates all user node of the target 'node'.
+    /// All user nodes are convolutions which satisfy options for onednn first conv using 'needs_onednn_bfyx_to_blocked'.
+    bool needs_all_usr_onednn_bfyx_to_blocked(format fmt_prev, const program_node& node);
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/include/layout_optimizer.h:214]: If it is used only within layout_optimizer, what about making it as static function?

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 7]
```diff
@@ -780,9 +759,33 @@ bool layout_optimizer::deconvolution_b_fs_yx_fsv16_opt(layout const &input_layou
...
+
+    return all_users_match;
+}
+
 bool layout_optimizer::needs_onednn_bfyx_to_blocked(format fmt_prev, format fmt_next, layout& prev_output_layout, const convolution_node& node) {
     auto next_output_layout = node.get_output_layout();
-    if (prev_output_layout.data_type != next_output_layout.data_type)
+    if (!(prev_output_layout.data_type == next_output_layout.data_type ||
+        (prev_output_layout.data_type == data_types::i8 && next_output_layout.data_type == data_types::u8) ||
+        (prev_output_layout.data_type == data_types::u8 && next_output_layout.data_type == data_types::i8)))
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:788]: This condition is difficult to understand.
Question: is `i8 -> u8` not supported in onednn?

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 8]
```diff
@@ -1400,8 +1404,6 @@ impl_types layout_optimizer::get_preferred_impl_type(program_node& node, format
                 (output_layout.format == format::b_fs_yx_fsv16) &&
                 !needs_onednn_bfyx_to_blocked(format::bfyx, output_layout.format, input_layout, conv))
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1405]: What does this condition mean? Can't we remove `first_conv` and `needs_onednn_bfyx_to_blocked` here?

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 9]
```diff
@@ -1578,6 +1572,10 @@ format layout_optimizer::get_preferred_format(program_node& node) {
         if (only_gemm_users(node)) {
             // TODO: Gemm is not supporting fsv layouts
             expected = format::bfyx;
+        } else if (use_onednn_impls && (layout.data_type == data_types::i8 || layout.data_type == data_types::u8) &&
+            needs_all_usr_onednn_bfyx_to_blocked(format::bfyx, node)) {
+            // All user nodes are convolutions which satisfy options for onednn first conv
+            expected = format::bs_fs_yx_bsv8_fsv4;
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1578]: I think this condition should be fused into existing condition. For example the one on line 1586. That is the previous condition for first conv..

**Review by isanghao — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gpu/src/graph/graph_optimizer/add_required_reorders.cpp (hunk 1 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/add_required_reorders.cpp:45]: As we had discussed, I resolved this issue at reorder_inputs step.
But I think it still has an issue. I will fix this by a separate PR.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp (hunk 2 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp:389]: Applied

File: src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp (hunk 3 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp:583]: Fixed

File: src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp (hunk 4 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp:585]: How about fixing them together at separate PR?

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp (hunk 7 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:788]: There is !

File: src/plugins/intel_gpu/src/graph/include/layout_optimizer.h (hunk 6 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/include/layout_optimizer.h:214]: I perfer this type cuz it is using needs_onednn_bfyx_to_blocked. Does it need to be fixed?

File: src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp (hunk 5 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp:589]: Applied

File: src/plugins/intel_gpu/src/graph/graph_optimizer/add_required_reorders.cpp (hunk 1 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/add_required_reorders.cpp:45]: Removed

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp (hunk 8 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp]: |    8 | 133                          | Convolution    |  2090 | [7]      | [1, 1, 128, 100]  | b_fs_yx_f32    | U8       | [1, 1, 128, 100]  | b_fs_yx_f32    | U8       | ocl:ref:any__u8                  |

Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp]: Modified

**Review by isanghao — APPROVED** [review]

**Pull request merged**